src/imitation/scripts/parallel.py

```diff
@@ -95,15 +126,114 @@ def parallel(
     )
 
     ray.init(**init_kwargs)
+    search_alg = search.Repeater(optuna.OptunaSearch(), repeat=repeat)
     try:
-        ray.tune.run(
-            trainable,
-            config=search_space,
-            name=run_name,
-            local_dir=local_dir,
-            resources_per_trial=resources_per_trial,
-            sync_config=ray.tune.syncer.SyncConfig(upload_dir=upload_dir),
-        )
+        if experiment_checkpoint_path:
+            if resume:
+                # restart failed runs from experiment_checkpoint_path
+                register_trainable("inner", trainable)
+                runner = ray.tune.execution.trial_runner.TrialRunner(
+                    local_checkpoint_dir=experiment_checkpoint_path,
+                    sync_config=ray.tune.syncer.SyncConfig(
+                        upload_dir=upload_dir,
+                        syncer=syncer,
+                    ),
+                    metric="mean_return",
+                    resume=resume,
+                )
+                print(
+                    "Live trials:",
+                    len(runner._live_trials),
+                    "/",
+                    len(runner._trials),
+                )
+                while not runner.is_finished():
+                    runner.step()
+                    print("Debug:", runner.debug_string())
+
+            # load experiment analysis results
+            result = ray.tune.ExperimentAnalysis(experiment_checkpoint_path)
+            result._load_checkpoints_from_latest(
+                glob.glob(experiment_checkpoint_path + "/experiment_state*.json"),
+            )
+            # update result.trials using all the experiment_state json files
+            result.trials = None
+            result.fetch_trial_dataframes()
+        else:
+            # run hyperparameter tuning
+            result = ray.tune.run(
+                trainable,
+                config=search_space,
+                num_samples=num_samples * repeat,
+                name=run_name,
+                local_dir=local_dir,
+                resources_per_trial=resources_per_trial,
+                sync_config=ray.tune.syncer.SyncConfig(
+                    upload_dir=upload_dir,
+                    syncer=syncer,
+                ),
+                search_alg=search_alg,
+                metric="mean_return",
+                mode="max",
+            )
+        if sacred_ex_name == "train_rl":
+            key_prefix = ""
+        elif sacred_ex_name == "train_preference_comparisons":
+            key_prefix = "rollout/"
+        else:
+            key_prefix = "imit_stats/"
+        key = key_prefix + "monitor_return_mean"
+
+        if eval_best_trial:
```

Comment:
After looking at this again I would have a strong preference to exclude the result evaluation from the `parallel.py` script. This scrip is already way to complex and grown over the years. Now we add another feature to it and I fear it will become entirely incomprehensible in the future. I am happy to discuss other solutions.